v-next/hardhat/src/internal/builtin-plugins/node/json-rpc/handler.ts

```diff
@@ -0,0 +1,294 @@
+      // or the current websocket connection isn't interested in the particular subscription.
+      if (isClosed || !subscriptions.includes(payload.subscription)) {
+        return;
+      }
+
+      try {
+        ws.send(
+          JSON.stringify({
+            jsonrpc: "2.0",
+            method: "eth_subscription",
+            params: payload,
+          }),
+        );
+      } catch (error) {
+        _handleError(error);
+      }
+    };
+
+    // Handle eth_subscribe notifications.
+    this.#provider.addListener("notification", listener);
+
+    ws.on("message", async (msg: string) => {
+      let rpcReq: JsonRpcRequest | JsonRpcRequest[];
+      let rpcResp: JsonRpcResponse | JsonRpcResponse[];
+
+      try {
+        rpcReq = _readWsRequest(msg);
+
+        rpcResp = Array.isArray(rpcReq)
+          ? await Promise.all(
+              rpcReq.map((req) =>
+                this.#handleSingleWsRequest(req, subscriptions),
+              ),
+            )
+          : await this.#handleSingleWsRequest(rpcReq, subscriptions);
+      } catch (error) {
+        rpcResp = _handleError(error);
+      }
+
+      ws.send(JSON.stringify(rpcResp));
+    });
+
+    ws.on("close", () => {
+      // Remove eth_subscribe listener.
+      this.#provider.removeListener("notification", listener);
+
+      // Clear any active subscriptions for the closed websocket connection.
+      isClosed = true;
+      subscriptions.forEach(async (subscriptionId) => {
+        await this.#provider.request({
+          method: "eth_unsubscribe",
+          params: [subscriptionId],
+        });
+      });
+    });
+  };
+
+  #sendEmptyResponse(res: ServerResponse) {
+    res.writeHead(200);
+    res.end();
+  }
+
+  #setCorsHeaders(res: ServerResponse) {
+    res.setHeader("Access-Control-Allow-Origin", "*");
+    res.setHeader("Access-Control-Request-Method", "*");
+    res.setHeader("Access-Control-Allow-Methods", "OPTIONS, GET");
+    res.setHeader("Access-Control-Allow-Headers", "*");
+  }
+
+  #sendResponse(
+    res: ServerResponse,
+    rpcResp: JsonRpcResponse | JsonRpcResponse[],
+  ) {
+    res.statusCode = 200;
+    res.setHeader("Content-Type", "application/json");
+    res.end(JSON.stringify(rpcResp));
+  }
+
+  async #handleSingleRequest(req: JsonRpcRequest): Promise<JsonRpcResponse> {
+    if (!isJsonRpcRequest(req)) {
+      return _handleError(new InvalidRequestError("Invalid request"));
+    }
+
+    const rpcReq: JsonRpcRequest = req;
+    let rpcResp: JsonRpcResponse | undefined;
+
+    try {
+      rpcResp = await this.#handleRequest(rpcReq);
+    } catch (error) {
+      rpcResp = _handleError(error);
+    }
+
+    // Validate the RPC response.
+    if (!isJsonRpcResponse(rpcResp)) {
+      // Malformed response coming from the provider, report to user as an internal error.
+      rpcResp = _handleError(new InternalError("Internal error"));
+    }
+
+    if (rpcReq !== undefined) {
```

Comment:
Isn’t this guaranteed to be non-undefined due to the earlier `isJsonRpcRequest` check?